ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/CommandGenerator.scala

```diff
@@ -60,24 +97,137 @@ final class CommandGenerator(
+        }
+        Seq.fill[String](f)(randomPayload(c.payloadSizeBytes))
+      }
+    val (templateDesc, createCmd) = templateName match {
+      case "Foo1" => (TemplateDescriptor.Foo1, Foo1(signatory, observers, payload).create.command)
+      case "Foo2" => (TemplateDescriptor.Foo2, Foo2(signatory, observers, payload).create.command)
+      case "Foo3" => (TemplateDescriptor.Foo3, Foo3(signatory, observers, payload).create.command)
+      case invalid => sys.error(s"Invalid template: $invalid")
     }
 
+    CreateCmdAndContinuations(
+      createCommand = createCmd,
+      continuationF = {
+        def createCont(cid: ContractId): Seq[Command] = {
+          val nonconsumingExercises = nonconsumingExercisePayload.map { payload =>
+            createExerciseCmd(
+              templateId = templateDesc.templateId,
+              choiceName = templateDesc.nonconsumingChoiceName,
+              argValue = payload,
+            )(cid)
+          }
+          val consumingExerciseO = consumingExercisePayload.fold[Option[Command]](None)(payload =>
+            Some(
+              createExerciseCmd(
+                templateId = templateDesc.templateId,
+                choiceName = templateDesc.consumingChoiceName,
+                argValue = payload,
+              )(cid)
+            )
+          )
+          // NOTE: Archive will be generated only if prior consuming exercise is absent.
+          val archiveExerciseO = if (archive && consumingExerciseO.isEmpty) {
+            Some(
+              doCreateArchiveExerciseCmd(
+                templateId = templateDesc.templateId,
+                cid = cid,
+              )
+            )
+          } else
+            None
+          nonconsumingExercises ++ consumingExerciseO.toList ++ archiveExerciseO.toList
+        }
+        createCont
+      },
+    )
+  }
+
+  private def createExerciseCmd(templateId: Identifier, choiceName: String, argValue: String)(
+      cid: ContractId
+  ): Command =
+    doCreateExerciseCmd(
+      templateId = templateId,
+      choiceName = choiceName,
+      argName = "exercisePayload",
+      argValue = argValue,
+      cid = cid,
+    )
+
+  private def doCreateExerciseCmd(
+      templateId: Identifier,
+      choiceName: String,
+      argName: String,
+      argValue: String,
+      cid: ContractId,
+  ): Command = {
+    val c: Command = Command(
+      command = Command.Command.Exercise(
+        value = ExerciseCommand(
+          templateId = Some(templateId),
+          contractId = cid.coid,
+          choice = choiceName,
+          choiceArgument = Some(
+            Value(
+              Value.Sum.Record(
+                Record(
+                  None,
+                  Seq(
+                    RecordField(
+                      label = argName,
+                      value = Some(Value(Value.Sum.Text(argValue))),
+                    )
+                  ),
+                )
+              )
+            )
+          ),
+        )
```

Comment:
nesting could have been reduced by introducing intermediate `val`s

Reply:
This feels natural to me as it's building a relatively simple AST. 
But not having a strong opinion on this though

Reply:
Introduced a val